File: problems/784_letter_case_permutation/solution.py
from typing import List


class Solution:
    def letterCasePermutation(self, s: str) -> List[str]:
        result = [[]]
        for c in s:
            if c.isalpha():
                result = [r + [v] for r in result for v in [c.lower(), c.upper()]]
            else:
                result = [r + [c] for r in result]
        return [''.join(r) for r in result]

File: problems/784_letter_case_permutation/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <set>
#include <map>
using namespace std;

class Solution {
public:
    vector<string> letterCasePermutation(string s) {
        vector<string> result = {""};
        for (char c : s) {
            vector<string> next;
            for (auto& r : result) {
                if (isalpha(c)) {
                    next.push_back(r + (char)tolower(c));
                    next.push_back(r + (char)toupper(c));
                } else {
                    next.push_back(r + c);
                }
            }
            result = next;
        }
        return result;
    }
};
